DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/compute_units/src/compute_units.hpp

```diff
@@ -0,0 +1,37 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#include <CL/sycl.hpp>
+#include <utility>
+
+namespace {
+template <class Func, template <std::size_t> class Name, std::size_t index>
+class SubmitOneComputeUnit {
+public:
+  SubmitOneComputeUnit(Func &&f, sycl::queue &q) {
+
+    q.submit([&](sycl::handler &h) {
+      h.single_task<Name<index>>([=] {
+        // verifies that f only takes a single argument
+        f(std::integral_constant<std::size_t, index>());
```

Comment:
I am puzzled by this code... What are you trying to do here?
Do you mean
```c++
f(index);
```
or this is an obfuscation contest?

Reply:
Operand needs to be a constant. However presumably `index` is already a constant since its a template parameter...

Reply:
Ok, I tried `f(index)` but it doesn't compile. Good to jog my memory on why all of this was necessary! I will try to give my own explanation here, but this reference might be more helpful: https://stackoverflow.com/questions/56937863/constexpr-lambda-argument

Yes, `index` is constexpr. However, there is actually no way for a function argument to be constexpr. So within the function `f`, even though it was passed a constexpr, in `f` it won't be known to be constexpr.

To get around this limitation, it's necessary to encode the value in a type instead. Then within `f` it's possible to query that type and obtain the value.

`std::integral_constant` is a way to encode a value in a type.

Reply:
I have found the reason of our disagreement and now I understand the reason for your code. :-)
When I saw old C++ coding style
```c++
         f(std::integral_constant<std::size_t, index>());
```
I actually parsed it:
```c++
         f(std::integral_constant<std::size_t, index>()());
```
so calling `std::integral_constant::operator()` returning the value itself, which is `index` here, according to https://en.cppreference.com/w/cpp/types/integral_constant.

If you had written instead a more modern
```c++
         f(std::integral_constant<std::size_t, index> {});
```
it would have seen immediately it was just creating a `std::integral_constant` and passing it.
The other super tricky part is also later
```c++
       auto f = Pipes::PipeAt<ID>::read();
```
which works because there is an implicit conversion operator `std::integral_constant::operator value_type` returning back a `std::size_t` from `ID`.
Ahem... All this without a single comment in a tutorial? How is a SYCL programmer expected to understand this? :-( This is the most complex code I have seen in oneAPI DPC++ so far... :-)
Another simpler way to do this nowadays could be to use a C++20 templated lambda where you put `index` directly in the template parameter. But actually the code from the end-user point of view would be longer. So your solution seems better!

Reply:
I'm glad we're on the same page! I will apply your tip about using an initializer list to make this more readable.

The tutorials should focus on teaching people how to use SYCL for FPGAs, not how C++ template metaprogramming works. We plan to move these header files into a separate library, so they can be re-used and the implementation details can be ignored. Nevertheless, more comments will also improve the quality of these libraries :-)

Reply:
> using an initializer list

No, in this case this is a uniform initialization, not an initializer list, even if both have `{}`. https://isocpp.org/wiki/faq/cpp11-language#uniform-init

Otherwise, I do not think this is ethical to push uncommented open-source code all over the place... I hate write-only code! :-)
If the code is open-source this is to learn from it...